test-framework/junit5-internal/src/main/java/io/quarkus/test/QuarkusProdModeTest.java

```diff
@@ -330,28 +323,48 @@ public Integer getExitCode() {
     private void exportArchive(Path deploymentDir, Class<?> testClass) {
         try {
             JavaArchive archive = getArchiveProducerOrDefault();
-            if (customApplicationProperties != null) {
-                archive.add(new PropertiesAsset(customApplicationProperties), "application.properties");
+
+            if (configResourceName != null) {
+                archive.addAsResource(configResourceName, APPLICATION_PROPERTIES);
             }
-            archive.as(ExplodedExporter.class).exportExplodedInto(deploymentDir.toFile());
 
-            String exportPath = System.getProperty("quarkus.deploymentExportPath");
-            if (exportPath != null) {
-                File exportDir = new File(exportPath);
-                if (exportDir.exists()) {
-                    if (!exportDir.isDirectory()) {
-                        throw new IllegalStateException("Export path is not a directory: " + exportPath);
+            if (customApplicationProperties != null) {
+                Node applicationProperties = archive.get(APPLICATION_PROPERTIES);
+                if (applicationProperties != null) {
+                    // We need to merge the existing "application.properties" asset and the overriden config properties
+                    Properties mergedProperties = new Properties();
+                    Asset asset = applicationProperties.getAsset();
+                    if (asset instanceof StringAsset strAsset) {
+                        mergedProperties.load(new StringReader(strAsset.getSource()));
+                    } else {
+                        try (InputStream in = asset.openStream()) {
+                            mergedProperties.load(in);
+                        }
                     }
-                    try (Stream<Path> stream = Files.walk(exportDir.toPath())) {
-                        stream.sorted(Comparator.reverseOrder()).map(Path::toFile)
-                                .forEach(File::delete);
+                    // overrideConfigKey() takes precedence
+                    customApplicationProperties.forEach(mergedProperties::put);
+
+                    if (Boolean.parseBoolean(System.getProperty("quarkus.test.log-merged-properties"))) {
+                        System.out.println("Merged config properties:\n" + mergedProperties.entrySet().stream()
```

Comment:
I'd use `keySet()` to be extra sure we're not logging secrets in values... The risk is low, but exists, e.g. if a module configures Maven to replace some properties in `application.properties` with an environment variable. Unlikely, I know; still.

Reply:
Good point.